### PR TITLE
Support creating sorted table

### PR DESCRIPTION
**Pull request by shixuan-fan**
- Allow creating and showing sorted table.
- Put sorted info into `HiveBucketProperty`.
- Still disallow inserting into sorted table.

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/Storage.java [hunk 1]
```diff
@@ -68,12 +68,6 @@ public String getLocation()
         return bucketProperty;
     }
 
-    @JsonProperty
-    public boolean isSorted()
-    {
-        return sorted;
```
Comment (kokosing) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/Storage.java]: do not you need to remove the field as well?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 2]
```diff
@@ -1589,6 +1656,27 @@ public void testShowCreateTable()
...
+                        "   bucketed_by = ARRAY['c4','c5'],\n" +
+                        "   format = 'RCBINARY',\n" +
+                        "   sorted_by = ARRAY['+c1','-c2']\n" +
+                        ")",
+                getSession().getCatalog().get(),
+                getSession().getSchema().get(),
+                "\"test_show_create_table'3\"");
+        assertUpdate(createTableSql);
+        actualResult = computeActual("SHOW CREATE TABLE \"test_show_create_table'3\"");
+        assertEquals(getOnlyElement(actualResult.getOnlyColumnAsSet()), createTableSql);
```
Comment (kokosing) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: it looks like you are missing `DROP TABLE` in all the tests

**Comment by kokosing** [comment]
Notice also travis failures.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSortOrder.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+public class HiveSortOrder
+{
+    private final String columnName;
+    private final int order;
+
+    @JsonCreator
+    public HiveSortOrder(@JsonProperty("column") String columnName, @JsonProperty("order") int order)
+    {
+        this.columnName = columnName;
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSortOrder.java]: can this be null?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSortOrder.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        this.order = order;
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSortOrder.java]: what are possible values here? From skimming the code it looks like `0` and `1`. Use enum for that.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java [hunk 5]
```diff
@@ -88,6 +92,27 @@ public HiveTableProperties(TypeManager typeManager, HiveClientConfig config)
...
+                        "sorting columns",
+                        typeManager.getType(parseTypeSignature("array(varchar)")),
+                        List.class,
+                        ImmutableList.of(),
+                        false,
+                        value -> ((Collection<?>) value).stream()
+                                .map(name -> {
+                                    checkState(
+                                            ((String) name).charAt(0) == '+' || ((String) name).charAt(0) == '-',
+                                            "Expect +/- to specify ASC/DESC for sorting columns");
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: `s/Expect to specify/Expected to denote` ?
Why not use `<col> ASC` here? Semantics would be more obvious.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java [hunk 6]
```diff
@@ -88,6 +92,27 @@ public HiveTableProperties(TypeManager typeManager, HiveClientConfig config)
...
+                        SORTED_BY_PROPERTY,
+                        "sorting columns",
+                        typeManager.getType(parseTypeSignature("array(varchar)")),
+                        List.class,
+                        ImmutableList.of(),
+                        false,
+                        value -> ((Collection<?>) value).stream()
+                                .map(name -> {
+                                    checkState(
+                                            ((String) name).charAt(0) == '+' || ((String) name).charAt(0) == '-',
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: You do the `((String) name)` cast 4 times here. Extract cast to separate `.map(String.class::cast)` step.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java [hunk 7]
```diff
@@ -88,6 +92,27 @@ public HiveTableProperties(TypeManager typeManager, HiveClientConfig config)
...
+                                            "Expect +/- to specify ASC/DESC for sorting columns");
+                                    return new HiveSortOrder(
+                                            ((String) name).substring(1).toLowerCase(ENGLISH),
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: Can `((String) name).substring(1).toLowerCase(ENGLISH)` contain spaces?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java [hunk 8]
```diff
@@ -132,8 +157,12 @@ public static HiveStorageFormat getHiveStorageFormat(Map<String, Object> tablePr
     public static Optional<HiveBucketProperty> getBucketProperty(Map<String, Object> tableProperties)
     {
         List<String> bucketedBy = getBucketedBy(tableProperties);
+        List<HiveSortOrder> sortedBy = getSortedBy(tableProperties);
         int bucketCount = (Integer) tableProperties.get(BUCKET_COUNT_PROPERTY);
         if ((bucketedBy.isEmpty()) && (bucketCount == 0)) {
+            if (!sortedBy.isEmpty()) {
+                throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s could be specified only when %s is specified", SORTED_BY_PROPERTY, BUCKETED_BY_PROPERTY));
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: Why?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 9]
```diff
@@ -596,6 +596,40 @@ private void testCreatePartitionedBucketedTableAsFewRows(Session session, HiveSt
...
+        String tableName = "test_create_bucketed_sorted_table_as";
+        @Language("SQL") String createTable = "" +
+                "CREATE TABLE " + tableName + " " +
+                "WITH (" +
+                "bucketed_by = ARRAY[ 'custkey', 'custkey2' ], " +
+                "bucket_count = 11, " +
+                "sorted_by = ARRAY[ '+custKey' ]" +
+                ") " +
+                "AS " +
+                "SELECT custkey, custkey AS custkey2, comment, orderstatus " +
```
Comment (findepi) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Put AS SELECT on single line

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 10]
```diff
@@ -596,6 +596,40 @@ private void testCreatePartitionedBucketedTableAsFewRows(Session session, HiveSt
...
+                "bucketed_by = ARRAY[ 'id' ], " +
+                "bucket_count = 11, " +
+                "sorted_by = ARRAY[ '+name' ]" +
+                ")";
+        assertUpdate(createTable);
+    }
+
+    @Test(
+            expectedExceptions = RuntimeException.class,
+            expectedExceptionsMessageRegExp = "Sorted tables cannot be created using CREATE TABLE AS")
```
Comment (findepi) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Add a test for CREATE TABLE AS .. WITH NO DATA.
Should succeed.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 11]
```diff
@@ -747,6 +781,39 @@ private void testCreateInvalidBucketedTable(HiveStorageFormat storageFormat)
...
+                    "INSERT INTO " + tableName + " " +
+                            "VALUES " +
+                            "  (1, VARCHAR 'b', VARCHAR 'c'), " +
+                            "  (2, 'bb', 'cc'), " +
+                            "  (3, 'bbb', 'ccc')",
+                    3);
+            fail("expected failure");
+        }
+        catch (Exception e) {
+            assertEquals(e.getMessage(), format("Inserting into bucketed sorted tables is not supported. Table 'tpch.%s'", tableName));
```
Comment (findepi) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: If the message is not what you expect (e.g. NPE with null message), the exception raised by the test will be hardly useful. Use `org.assertj.core.api.Assertions#assertThatThrownBy`:
```
assertThatThrownBy(() -> assertUpdate(...))
  .hasMessage(...)
```

File: presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestMetastoreUtil.java [hunk 12]
```diff
@@ -143,18 +143,4 @@ public void testHiveSchemaPartition()
         Properties actual = MetastoreUtil.getHiveSchema(ThriftMetastoreUtil.fromMetastoreApiPartition(TEST_PARTITION_WITH_UNSUPPORTED_FIELDS), ThriftMetastoreUtil.fromMetastoreApiTable(TEST_TABLE_WITH_UNSUPPORTED_FIELDS));
         assertEquals(actual, expected);
     }
-
-    @Test(expectedExceptions = IllegalArgumentException.class, expectedExceptionsMessageRegExp = "Writing to sorted and/or skewed table/partition is not supported")
```
Comment (findepi) [inline presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestMetastoreUtil.java]: Why is this removed?

File: presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java [hunk 13]
```diff
@@ -378,8 +380,8 @@ public static void fromMetastoreApiStorageDescriptor(StorageDescriptor storageDe
 
     private static StorageDescriptor makeStorageDescriptor(String tableName, List<Column> columns, Storage storage)
     {
-        if (storage.isSorted() || storage.isSkewed()) {
-            throw new IllegalArgumentException("Writing to sorted and/or skewed table/partition is not supported");
+        if (storage.isSkewed()) {
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreUtil.java]: Use `checkArgument`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 14]
```diff
@@ -596,6 +596,40 @@ private void testCreatePartitionedBucketedTableAsFewRows(Session session, HiveSt
         assertFalse(getQueryRunner().tableExists(session, tableName));
     }
 
+    @Test
+    public void testCreateBucketedSortedTable()
+    {
+        String tableName = "test_create_bucketed_sorted_table";
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: This is only used once, no need for a variable

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 15]
```diff
@@ -596,6 +596,40 @@ private void testCreatePartitionedBucketedTableAsFewRows(Session session, HiveSt
...
+                "(id int, name varchar, ds varchar) " +
+                "WITH (" +
+                "bucketed_by = ARRAY[ 'id' ], " +
+                "bucket_count = 11, " +
+                "sorted_by = ARRAY[ '+name' ]" +
+                ")";
+        assertUpdate(createTable);
+    }
+
+    @Test(
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Don't wrap these (none of the others do)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 16]
```diff
@@ -747,6 +781,39 @@ private void testCreateInvalidBucketedTable(HiveStorageFormat storageFormat)
...
+                "bucketed_by = ARRAY[ 'id' ], " +
+                "bucket_count = 11, " +
+                "sorted_by = ARRAY[ '+name' ]" +
+                ")";
+        assertUpdate(createTable);
+
+        try {
+            assertUpdate(
+                    "INSERT INTO " + tableName + " " +
+                            "VALUES " +
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: This can just be a single value

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 17]
```diff
@@ -596,6 +596,40 @@ private void testCreatePartitionedBucketedTableAsFewRows(Session session, HiveSt
         assertFalse(getQueryRunner().tableExists(session, tableName));
     }
 
+    @Test
+    public void testCreateBucketedSortedTable()
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: This test seems redundant with the one in `testShowCreateTable` and doesn't actually validate that we created a sorted table. Probably just remove it.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 18]
```diff
@@ -596,6 +596,40 @@ private void testCreatePartitionedBucketedTableAsFewRows(Session session, HiveSt
...
+                ")";
+        assertUpdate(createTable);
+    }
+
+    @Test(
+            expectedExceptions = RuntimeException.class,
+            expectedExceptionsMessageRegExp = "Sorted tables cannot be created using CREATE TABLE AS")
+    public void testCreateBucketedSortedTableAs()
+    {
+        String tableName = "test_create_bucketed_sorted_table_as";
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Inline

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 19]
```diff
@@ -596,6 +596,40 @@ private void testCreatePartitionedBucketedTableAsFewRows(Session session, HiveSt
...
+    public void testCreateBucketedSortedTable()
+    {
+        String tableName = "test_create_bucketed_sorted_table";
+        @Language("SQL") String createTable = "" +
+                "CREATE TABLE " + tableName + " " +
+                "(id int, name varchar, ds varchar) " +
+                "WITH (" +
+                "bucketed_by = ARRAY[ 'id' ], " +
+                "bucket_count = 11, " +
+                "sorted_by = ARRAY[ '+name' ]" +
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Don't put spaces around array brackets

File: presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestMetastoreUtil.java (hunk 12 above)
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/metastore/TestMetastoreUtil.java]: This should test that skewed tables fail

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java (hunk 10 above)
Comment (shixuan-fan) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Per discussion with @electrum , this will require a larger change in the engine, which is unrelated to this PR. We could tackle that later.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java (hunk 7 above)
Comment (findepi) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: Did you add some validation for this?

Comment (shixuan-fan) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java]: I'll add them to the place similar to bucketing column check. Thanks for the catch!

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java [hunk 20]
```diff
@@ -51,7 +58,13 @@ public HiveBucketProperty(
         if (!bucketColsSet) {
             throw new PrestoException(HIVE_INVALID_METADATA, "Table/partition metadata has 'numBuckets' set, but 'bucketCols' is not set: " + tablePartitionName);
         }
-        return Optional.of(new HiveBucketProperty(storageDescriptor.getBucketCols(), storageDescriptor.getNumBuckets()));
+        checkState(
+                storageDescriptor.getSortCols().stream().allMatch(sortCol -> sortCol.getOrder() == 1 || sortCol.getOrder() == 0),
+                "Expect only 0 or 1 for elements of sortCols in storageDescriptor");
```
[the remark this answers is not in the document]
Comment (shixuan-fan) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java]: How about `Order in sorting columns can only be 0 or 1`?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSortOrder.java [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class HiveSortOrder
+{
+    public enum Ordering
```
[the remark this answers is not in the document]
Comment (shixuan-fan) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveSortOrder.java]: Because we also need to store column names, which I think is beyond what an enum should do. However, now I feel `HiveSortOrder` is kind of confusing. I chose this name because this class is behave like `org.apache.hadoop.hive.metastore.api.Order`, but perhaps we should name this class `SortingColumn` to avoid confusion?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java [hunk 22]
```diff
@@ -77,7 +96,8 @@ public boolean equals(Object o)
         }
         HiveBucketProperty that = (HiveBucketProperty) o;
         return bucketCount == that.bucketCount &&
-                Objects.equals(bucketedBy, that.bucketedBy);
+                Objects.equals(bucketedBy, that.bucketedBy) &&
+                Objects.equals(sortedBy, that.sortedBy);
```
[the remark this answers is not in the document]
Comment (shixuan-fan) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketProperty.java]: Fixed.

**Comment by shixuan-fan** [comment]
Ping @electrum

**Comment by shixuan-fan** [comment]
Superseded by #10608
